src/components/application_manager/src/resumption/resumption_data_processor.cc

```diff
@@ -0,0 +1,1069 @@
+      !global_properties[strings::keyboard_properties].empty() ||
+      !global_properties[strings::menu_icon].empty() ||
+      !global_properties[strings::menu_title].empty() ||
+      !global_properties[strings::timeout_prompt].empty() ||
+      !global_properties[strings::vr_help].empty() ||
+      !global_properties[strings::vr_help_title].empty();
+
+  LOG4CXX_DEBUG(logger_,
+                std::boolalpha
+                    << "Application has global properties to restore: "
+                    << has_gl_props_to_restore);
+
+  return has_gl_props_to_restore;
+}
+
+bool ResumptionDataProcessor::HasSubscriptionsToRestore(
+    const smart_objects::SmartObject& saved_app) const {
+  LOG4CXX_AUTO_TRACE(logger_);
+
+  const smart_objects::SmartObject& subscriptions =
+      saved_app[strings::application_subscriptions];
+
+  const bool has_ivi_subscriptions =
+      !subscriptions[strings::application_vehicle_info].empty();
+
+  const bool has_button_subscriptions =
+      !(subscriptions[strings::application_buttons].length() == 1 &&
+        static_cast<hmi_apis::Common_ButtonName::eType>(
+            subscriptions[strings::application_buttons][0].asInt()) ==
+            hmi_apis::Common_ButtonName::CUSTOM_BUTTON);
+
+  const bool has_waypoints_subscriptions =
+      subscriptions[strings::subscribed_for_way_points].asBool();
+
+  bool has_appservice_subscriptions = false;
+  const char* app_service_info = "appService";
+  if (subscriptions.keyExists(app_service_info)) {
+    has_appservice_subscriptions = !subscriptions[app_service_info].empty();
+  }
+
+  bool has_system_capability_subscriptions = false;
+  const char* application_system_capability = "systemCapability";
+  if (subscriptions.keyExists(application_system_capability)) {
+    has_system_capability_subscriptions =
+        !subscriptions[application_system_capability].empty();
+  }
+
+  const bool has_subscriptions_to_restore =
+      has_ivi_subscriptions || has_button_subscriptions ||
+      has_waypoints_subscriptions || has_appservice_subscriptions ||
+      has_system_capability_subscriptions;
+
+  LOG4CXX_DEBUG(logger_,
+                std::boolalpha << "Application has subscriptions to restore: "
+                               << has_subscriptions_to_restore);
+
+  return has_subscriptions_to_restore;
+}
+
+bool ResumptionRequestIDs::operator<(const ResumptionRequestIDs& other) const {
+  return correlation_id < other.correlation_id ||
+         function_id < other.function_id;
+}
+
+void ResumptionDataProcessor::ProcessResponseFromHMI(
+    const smart_objects::SmartObject& response,
+    const hmi_apis::FunctionID::eType function_id,
+    const int32_t corr_id) {
+  ResumptionRequestIDs request_ids;
+  request_ids.function_id = function_id;
+  request_ids.correlation_id = corr_id;
+  // TODO i suppose it can be optimised with moving app id into
+  // ResumptionRequest struct, so that we don't have to perform
+  // basically the same search twice
```

Comment:
I've explored this suggestion and don't find how it may work or help. I think we need to remove this ToDo as misleading and create separate function for finding app id on event.

Reply:
Also deleted in 63d3bed188bbfd104a15306cf0086c677c4fa8cc